src/HappyEyeBallsConnectionBuilder.php

```diff
@@ -0,0 +1,288 @@
+    public $connectQueue = array();
+    public $timer;
+    public $parts;
+    public $ipsCount = 0;
+    public $failureCount = 0;
+    public $resolve;
+    public $reject;
+
+    public static function create(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver, $uri)
+    {
+        $connectionBuilder = new self($loop, $connector, $resolver, $uri);
+
+        return $connectionBuilder->connect();
+    }
+
+    private function __construct(LoopInterface $loop, ConnectorInterface $connector, Resolver $resolver, $uri)
+    {
+        $this->loop = $loop;
+        $this->connector = $connector;
+        $this->resolver = $resolver;
+        $this->uri = $uri;
+
+        if (\strpos($uri, '://') === false) {
+            $this->parts = \parse_url('tcp://' . $uri);
+            unset($this->parts['scheme']);
+        } else {
+            $this->parts = \parse_url($uri);
+        }
+
+        if (!$this->parts || !isset($this->parts['host'])) {
+            throw new \InvalidArgumentException('Given URI "' . $uri . '" is invalid');
+        }
+
+        $this->host = \trim($this->parts['host'], '[]');
+    }
+
+    public function connect()
+    {
+        // skip DNS lookup / URI manipulation if this URI already contains an IP
+        if (false !== \filter_var($this->host, \FILTER_VALIDATE_IP)) {
+            return $this->connector->connect($this->uri);
+        }
+
+        $that = $this;
+        return new Promise\Promise(function ($resolve, $reject) use ($that) {
+            $that->resolve = $resolve;
+            $that->reject = $reject;
+            foreach (array(Message::TYPE_AAAA, Message::TYPE_A) as $type) {
+                $that->resolved[$type] = false;
+                $that->resolverPromises[$type] = $that->resolver->resolveAll($that->host, $type)->then(function (array $ips) use ($that, $type) {
+                    unset($that->resolverPromises[$type]);
+                    $that->resolved[$type] = true;
+
+                    $that->pushIpsOntoQueue($ips);
+
+                    $that->check();
+                }, function () use ($type, $reject, $that) {
+                    unset($that->resolverPromises[$type]);
+                    $that->resolved[$type] = true;
+
+                    if ($that->hasBeenResolved() === false) {
+                        return;
+                    }
+
+                    if ($that->ipsCount === 0) {
+                        $that->resolved = null;
+                        $that->resolverPromises = null;
+                        $reject(new \RuntimeException('Connection to ' . $that->uri . ' failed during DNS lookup: DNS error'));
+                    }
+                });
+            }
+        }, function ($_, $reject) use ($that) {
+            $that->cleanUp();
+
+            $reject(new \RuntimeException('Connection to ' . $that->uri . ' cancelled during DNS lookup'));
+
+            $_ = $reject = null;
+        });
+    }
+
+    public function check()
```

Comment:
All public methods should probably be marked `@internal`?

Reply:
That would make sense signalling that those methods are internal for the class (which is internal on it's own).

Reply:
done